Mage/src/main/java/mage/util/CardUtil.java

```diff
@@ -1009,6 +1010,58 @@ public static Set<UUID> getAllPossibleTargets(Ability ability, Game game) {
                 .collect(Collectors.toSet());
     }
 
+    /**
+     * For finding the spell or ability on the stack for "becomes the target" triggers.
+     * @param event the GameEvent.EventType.TARGETED from checkTrigger()
+     * @param game the Game from checkTrigger()
+     * @return the StackObject which targeted the source, or null if not found
+     */
+    public static StackObject getTargetingStackObject(GameEvent event, Game game) {
+        // In case of multiple simultaneous triggered abilities from the same source,
+        // need to get the actual one that targeted, see #8026, #8378
+        // Also avoids triggering on cancelled selections, see #8802
+        for (StackObject stackObject : game.getStack()) {
+            Ability stackAbility = stackObject.getStackAbility();
+            if (stackAbility == null || !stackAbility.getSourceId().equals(event.getSourceId())) {
+                continue;
+            }
+            for (Target target : stackAbility.getTargets()) {
+                if (target.getTargets().contains(event.getTargetId())) {
+                    return stackObject;
+                }
+            }
+        }
+        return null;
+    }
+
+    /**
+     * For ensuring that spells/abilities that target the same object twice only trigger "becomes the target" abilities once
+     * @param triggerId this.id of the TriggeredAbility
+     * @param targetingObject from getTargetingStackObject
+     * @param event the GameEvent.EventType.TARGETED from checkTrigger()
+     * @param game the Game from checkTrigger()
+     * @return false if already triggered, true if this is the first/only trigger
+     */
+    public static boolean checkTargetMap(UUID triggerId, StackObject targetingObject, GameEvent event, Game game) {
+        // If a spell or ability an opponent controls targets a single permanent you control more than once,
+        // Battle Mammoth's triggered ability will trigger only once.
+        // However, if a spell or ability an opponent controls targets multiple permanents you control,
+        // Battle Mammoth's triggered ability will trigger once for each of those permanents. (2021-02-05)
+        Map<UUID, Set<UUID>> targetMap = (Map<UUID, Set<UUID>>) game.getState().getValue("targetMap" + triggerId);
+        // targetMap: key - targetId; value - Set of stackObject Ids
+        if (targetMap == null) {
+            targetMap = new HashMap<>();
+        }
+        Set<UUID> targetingObjects = targetMap.computeIfAbsent(event.getTargetId(), k -> new HashSet<>());
+        if (!targetingObjects.add(targetingObject.getId())) {
+            return false; // The triggered ability already recorded that target of the stack object
+        }
+        // Otherwise, store this combination of triggered ability / target / stack object
+        targetMap.put(event.getTargetId(), targetingObjects);
+        game.getState().setValue("targetMap" + triggerId, targetMap);
```

Comment:
Its looks bugged with AI games (with any games simulation usage). Game state value must be immutable (copied with all values instead object ref). Object ref can be affected by game simulations bug - game simulation can use object ref to change data of the real game.

Also must use names like targetedMap and checkTargetedTriggerAlreadyUsed/checkTargetedTriggerNotUsed or like that.

Reply:
Sorry not quite sure what you mean here:
1. Need to check `isSimulation()` to not check target map for during game simulation?
2. Need to store something other than the map in the game state value, to prevent anything persisting through a rollback?

Original idea was implemented in back in cd4b93a7. It's only for use in the moment of targeting- my understanding is that targeting objects get fresh UUID when put on the stack. But perhaps good practice anyway to store unique game state value keys for each combination, with just a boolean in the end. I can rework it. And I can improve the naming.

Reply:
You must save new targetMap on change, e.g. targetMap = new HashMap(targetMap). It will help with object ref bugs